### PR TITLE
Support Jenkins CI

### PR DESCRIPTION
**Pull request by AbdYsn**
This patch adds a Jenkins file example, it also modifies the E2E
kind script to support a new mode of operation: the bash service.
The modification was done to support running the scripts in a limited
privilege shell.

### CODE REVIEW

**Comment by AbdYsn** [comment]
Another jenkins job is configured to update the admin list. It needs to be run from a PR and can be triggered using `/update-admins`

**Comment by AbdYsn** [comment]
@martinkennelly @adrianchiris can you guys PTAL, thanks.

**Comment by adrianchiris** [comment]
Thanks @AbdYsn ! will review this soon

File: ci/jenkins/sriov-network-operator-ci.yaml
Comment (adrianchiris) [reply ci/jenkins/sriov-network-operator-ci.yaml:1]: could you also add comments in the parts that are expected to be changes when incorporating this into an existing jenkins.

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:67]: > "No such option!!"

It reads abit aggressive :) , maybe just :

echo "Error: Invalid option, Exiting..."

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:36]: can you test if file exists and exit otherwise  ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:26]: options (flags) usually have no order.

if every PF needs to have a NS associated, then you can provide them in the same option e.g `pf:ns`

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:13]: nit: make this configurable like timeout ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:90]: did you mean to log an error here ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:132]: can you prefix all errors with either `Error:` or `ERROR:`  ? (pick one and stick with it :) )

Comment (adrianchiris) [reply hack/vf-netns-switcher.sh:132]: maybe add some helper funcs 

`log_error_exit()`
`log_error()`
`log()`

and use them throughout the code ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:304]: will we ever reach this point ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:318]: unless i missed something, status is first defined here why do not use `$?` directly ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:306]: can you add some logs on the different stages performed here to provide some high level info for the user ?

**Review by adrianchiris — COMMENTED** [review]
Thanks for submitting this !

I still need to give `vf-netns-switcher.sh` a deeper look. provided some comments/questions/nits so we can get the ball rolling :)

im thinking after this we should split the kind cluster setup from tests, this will allow to reuse this for development environment setup as well.

File: hack/e2e.md
Comment (AbdYsn) [reply hack/e2e.md]: used yq since we already use it in our CIs, but thinking about it, moving to jq would be better in this case.
Made the change :)

File: hack/e2e.md
Comment (AbdYsn) [reply hack/e2e.md]: done and moved to `doc/testing-kind.md`

File: hack/e2e.md
Comment (AbdYsn) [reply hack/e2e.md]: done

File: hack/e2e.md
Comment (AbdYsn) [reply hack/e2e.md]: used `test-suit`, WDYT? will fix the typo, just realized it.

File: hack/e2e.md
Comment (AbdYsn) [reply hack/e2e.md]: used `system-service`

File: hack/e2e.md
Comment (AbdYsn) [reply hack/e2e.md]: done

File: hack/e2e.md
Comment (AbdYsn) [reply hack/e2e.md]: The main reason for that is the ability to run the script as none root with limited privileges. since the service script needs full root privilege, only root should be able to edit it and use it.

File: ci/admin-list.yaml
Comment (AbdYsn) [reply ci/admin-list.yaml:1]: added a simple README, please have a look,
currently it is only jenkins that will use the file, it will be used by a jenkins job to update who can trigger CI.

File: ci/jenkins/sriov-network-operator-ci.yaml
Comment (AbdYsn) [reply ci/jenkins/sriov-network-operator-ci.yaml]: it is the exact same, since the file should be the source of this.

File: ci/jenkins/sriov-network-operator-ci.yaml
Comment (AbdYsn) [reply ci/jenkins/sriov-network-operator-ci.yaml]: added it.

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh]: thanks, done

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh:50]: done

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh]: done, using `--netns` flag

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh]: done

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh:115]: no, it is used as a return value for the function. like `return null`.

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh]: prefixed the error with Error, i dont think we will need to add a wrapper function, since as far as i can tell the file will not change alot.

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh]: no, removed it.

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh]: you are correct, done out of habit, fixed it.

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh:332]: added them to the corresponding functions.

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:118]: nit: add `return 0` for consistency ?

although return value is not being used IIUC so you could just use plain `return` in all cases here

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:368]: is it desirable to have a trap so that when the service exits it will move netdevs back to default namespace ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:365]: i didnt see variable pf defined in this scope. can you point me to it ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:216]: siwitchdev -> switchdev

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:122]: why not just $1 ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:108]: why not just $2 ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:143]: why no just $1 ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:156]: why not just $2 ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:186]: why not just $2 ?

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [inline hack/vf-netns-switcher.sh:202]: same.

in general this to me is very odd as netns is defined as part of a for loop in the logic at the bottom and is never defined as a global variable at the top.

so one needs to search where this method is called.
also the second value is never used as this method is always called with an argument

File: hack/run-e2e-test-kind.sh
Comment (adrianchiris) [reply hack/run-e2e-test-kind.sh:28]: I would prefer to just have it as a flag like `device-netns-switcher` so we can easily extend in the future.
dont think backward compat is an issue here, worst case is minor update in test infra.

In any case, we can leave this as is and change later if needed.

File: hack/vf-netns-switcher.sh
Comment (adrianchiris) [reply hack/vf-netns-switcher.sh:370]: how about this:

1. create namespace with random name
2. move VFs from ns in config to this namespace
3. delete namespace from 1(will put netdev back in root namespace)

File: hack/run-e2e-test-kind.sh
Comment (adrianchiris) [inline hack/run-e2e-test-kind.sh:112]: need to make sure directory exists

Getting the following:

```
## make KinD's sysfs writable (required to create VFs)                                                                                                                                                           ## label KinD's control-plane-node as sriov capable                                                                                                                                                              node/kind-worker labeled                                                                                                                                                                                         ## label KinD worker as worker                                                                                                                                                                                   node/kind-worker labeled                                                                                                                                                                                         ./hack/run-e2e-test-kind.sh: line 112: /etc/vf-switcher/vf-switcher.yaml: No such file or directory 
```

File: doc/testing-kind.md
Comment (adrianchiris) [inline doc/testing-kind.md:68]: this is deleted in teardown-e2e-kind-cluster.sh so i dont think you need it here

File: doc/testing-kind.md
Comment (adrianchiris) [inline doc/testing-kind.md:69]: missing `rm /usr/bin/vf-netns-switcher.sh`

File: hack/run-e2e-test-kind.sh
Comment (adrianchiris) [inline hack/run-e2e-test-kind.sh:8]: suit -> suite

File: hack/run-e2e-test-kind.sh
Comment (adrianchiris) [inline hack/run-e2e-test-kind.sh:9]: suit -> suite

File: hack/run-e2e-test-kind.sh
Comment (AbdYsn) [reply hack/run-e2e-test-kind.sh:113]: Done

File: doc/testing-kind.md
Comment (AbdYsn) [reply doc/testing-kind.md]: removed

File: doc/testing-kind.md
Comment (AbdYsn) [reply doc/testing-kind.md:69]: added

File: hack/run-e2e-test-kind.sh
Comment (AbdYsn) [reply hack/run-e2e-test-kind.sh]: fixed

File: hack/run-e2e-test-kind.sh
Comment (AbdYsn) [reply hack/run-e2e-test-kind.sh]: fixed

File: hack/teardown-e2e-kind-cluster.sh
Comment (adrianchiris) [reply hack/teardown-e2e-kind-cluster.sh:10]: can you add a conditional check here on stopping the service ? if we use test-suite vf switcher this emits an error

File: hack/run-e2e-test-kind.sh
Comment (AbdYsn) [reply hack/run-e2e-test-kind.sh:35]: Added the flag, and also kept this part for backward compatibility since it will not affect functionality in any way, though the priority will be for what came later.

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh:405]: Done, can you have a look

File: ci/README.md
Comment (adrianchiris) [inline ci/README.md:2]: first time seeing "CIes", can you just use CI throughout this file (or CIs)?

Comment (AbdYsn) [reply ci/README.md:2]: done, changed it to `Vendors CI`

File: ci/README.md
Comment (adrianchiris) [inline ci/README.md:23]: there are a couple of things that are unclear to me.

can you elaborate on the expected CI behaviour when <sub-test> is not provided in test/skip and when both <vendor> and <sub-test> is not provided in test/skip ?

also what is the expected behaviour with skip in general ?

File: hack/teardown-e2e-kind-cluster.sh
Comment (adrianchiris) [inline hack/teardown-e2e-kind-cluster.sh:10]: This will not work if you use `--device-netns-switcher`

File: ci/README.md
Comment (AbdYsn) [reply ci/README.md:23]: Nothing would happen, essentially a job gets triggered on a specific trigger phrase, if you replace one section with `all` all the tests for that section will get triggered. 
To elaborate, consider a test called `e2e-nvidia-cx5`, if you want to test it, you should comment one of the following:
 *  `/test-e2e-nvidia-cx5`
 * `/test-e2e-nvidia-all`
 * `/test-e2e-all`
 * `/test-all`

For the skip, it would be a separate job which gets triggered on the `/skip` phrases and the same trigger convention as the `/test` convention above. The job would just `exit 0` to have a green report on the PR with the same `status-context` as the `test` job.

File: hack/teardown-e2e-kind-cluster.sh
Comment (AbdYsn) [reply hack/teardown-e2e-kind-cluster.sh]: done, can you have a look.

**Comment by adrianchiris** [comment]
testing mellanox CI:

/test-all

**Review by adrianchiris — APPROVED** [review]
CI triggers work. and Nvidia CI is running.

**Review by e0ne — APPROVED** [review]

**Review by zshi-redhat — COMMENTED** [review]
@AbdYsn Thanks for your continued effort on making vendor CI works! appreciate it!

**Comment by zshi-redhat** [comment]
> /test-e2e-nvidia-cx5

It seems the CI is not triggered, maybe I need to wait until this PR is merged so that admin list takes effect?

**Comment by zshi-redhat** [comment]
Is there a way to check the jenkins log? I didn't see the `Details` link in the `Nvidia e2e test CI` (maybe it is because the test is marked as skipped now)?

**Comment by Eoghan1232** [comment]
> /test-all

I assume as I am not on Admin list, I cannot run Jenkins job? 
@AbdYsn

File: ci/README.md
Comment (AbdYsn) [reply ci/README.md:41]: > 

Not quite, the patch should be analyze, and determined whether it should be tested or not against the vendor CI. If it should then the `/test` is used to trigger the job and merge will be blocked until the CI is green (either by retesting the test, or uploading a new patchset), ideally the CI would only fail because of a change in the patch and logs should be provided to be able to debug why the patch failed. 
Only when it is determined that there is no need to test this test against this patch, then the `/skip` should be used to pass the test.



> 

well it is not just `device`, it can also be other stuff, say `openshift`, `kind`, and `baremetal` can also be added as a `sub-test`. A note here is that another github workflow will be added after this patch is merged to comment available CI triggers on PR creation.

**Comment by AbdYsn** [comment]
> > /test-e2e-nvidia-cx5
> 
> It seems the CI is not triggered, maybe I need to wait until this PR is merged so that admin list takes effect?

That actually because there is no such trigger. currently because there is only a single test, and i did not want to tie it to CX5 devices, i did not add a subtest to job trigger, and it can be trigger using the `/test-e2e-nvidia-all` trigger. Also since the NVIDIA jenkins instance can not be accessed outside the company, the polling mechanism is used to trigger the jobs against the webhook mechanism. so there is a delay of at most 5 minutes until the trigger will take effet.

**Comment by AbdYsn** [comment]
> Is there a way to check the jenkins log? I didn't see the `Details` link in the `Nvidia e2e test CI` (maybe it is because the test is marked as skipped now)?

my mistake, forgot to enable them in the CI configuration, will do that shortly

**Comment by AbdYsn** [comment]
> update-

Correct, If you want to be added to the `admin-list`, then a patch should be proposed to add you to the admin list, and for you to actually be added, on the proposed patch a current admin should comment `/update-admins` and then merge the commit. A note here is that there is a daily job that would read the `main` branch admin-list and update the job admin list accordingly, so the admin list will be overridden once each day to align it with `main` branch. The other way to update it is to contact the maintainer of the verndor CI and ask him to update it manually.

**Comment by Eoghan1232** [comment]
> 

Ah, I see. 
Just to clarify.
Steps: 
PR -> adding user to the admins list. (admin-list.yaml)
Admin comments: '/update-admins' before it is merged.
Then merge after.

File: hack/vf-netns-switcher.sh
Comment (Eoghan1232) [inline hack/vf-netns-switcher.sh:216]: Should it return 0?

File: hack/vf-netns-switcher.sh
Comment (Eoghan1232) [inline hack/vf-netns-switcher.sh:233]: Same as above. Should it Return 0?

**Review by Eoghan1232 — COMMENTED** [review]
LGTM. 
Thank you for working on this!

**Comment by AbdYsn** [comment]
@zshi-redhat can you try triggering it now with any of the supported phrases?

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh:218]: you are correct, Done

File: hack/vf-netns-switcher.sh
Comment (AbdYsn) [reply hack/vf-netns-switcher.sh:237]: I think it should actually be returning 1 if it reached this point, added an else case to the outer if that would return 1, this way we will never reach outside the ifs.

**Comment by AbdYsn** [comment]
> > > > /test-all
> > > 
> > > 
> > > I assume as I am not on Admin list, I cannot run Jenkins job? @AbdYsn
> > 
> > 
> > Correct, If you want to be added to the `admin-list`, then a patch should be proposed to add you to the admin list, and for you to actually be added, on the proposed patch a current admin should comment `/update-admins` and then merge the commit. A note here is that there is a daily job that would read the `main` branch admin-list and update the job admin list accordingly, so the admin list will be overridden once each day to align it with `main` branch. The other way to update it is to contact the maintainer of the verndor CI and ask him to update it manually.
> 
> Ah, I see. Just to clarify. Steps: PR -> adding user to the admins list. (admin-list.yaml) Admin comments: '/update-admins' before it is merged. Then merge after.

Yes, if you want to have the admin list updated immediately.

File: ci/README.md
Comment (zshi-redhat) [reply ci/README.md:41]: I see, thanks for the explanation!
I think we don't define openshift, kind or baremetal yet, right? they are just possible examples, could be well be a different type (platform for example) in the test- convention.

Comment (AbdYsn) [reply ci/README.md:41]: > I think we don't define openshift, kind or baremetal yet, right?

right

> could be well be a different type (platform for example) in the test- convention

yes

**Review by zshi-redhat — APPROVED** [review]

**Comment by AbdYsn** [comment]
Checking the logs, /test-all

**Review by Eoghan1232 — APPROVED** [review]

**Comment by adrianchiris** [comment]
@AbdYsn BIG thanks for working on it ! 🚀

**Pull request merged**
